backend/src/providers/transactions.ts

```diff
@@ -126,6 +130,109 @@ export const getTransactionInfo = async (
   return transaction[0] || null;
 };
 
+export const getTransactionDetails = async (
+  transactionHash: string
+): Promise<TransactionDetails | null> => {
+  const transactionBaseInfo = await getTransactionInfo(transactionHash);
+
+  if (!transactionBaseInfo) {
+    throw new Error(`No hash ${transactionHash} found`);
+  }
+  const transactionInfo = await sendJsonRpc("EXPERIMENTAL_tx_status", [
+    transactionBaseInfo.hash,
+    transactionBaseInfo.signerId,
+  ]);
+  const blockHashes = transactionInfo.receipts_outcome.map(
+    (receipt) => receipt.block_hash
+  );
+  const blockHeights = await getBlockHeightsByHashes(blockHashes);
+  const includedInBlockMap = blockHeights.reduce((acc, block) => {
+    acc.set(block.blockHash, {
+      blockHeight: block.blockHeight,
+      blockTimestamp: block.blockTimestamp,
+    });
+    return acc;
+  }, new Map());
+  const receiptsOutcome = transactionInfo.receipts_outcome.map((receipt) => ({
+    id: receipt.id,
+    proof: receipt.proof,
+    outcome: receipt.outcome,
+    includedInBlock: {
+      hash: receipt.block_hash,
+      height: includedInBlockMap.get(receipt.block_hash).blockHeight,
+      timestamp: includedInBlockMap.get(receipt.block_hash).blockTimestamp,
+    },
+  }));
+
+  if (
+    transactionInfo.receipts.length === 0 ||
+    transactionInfo.receipts[0].receipt_id !== receiptsOutcome[0].id
+  ) {
+    transactionInfo.receipts.unshift({
+      predecessor_id: transactionInfo.transaction.signer_id,
+      receipt: {
+        Action: {
+          signer_id: transactionInfo.transaction.signer_id,
+          signer_public_key: "",
+          gas_price: "0",
+          output_data_receivers: [],
+          input_data_ids: [],
+          actions: transactionInfo.transaction.actions,
+        },
+      },
+      receipt_id: receiptsOutcome[0].id,
+      receiver_id: transactionInfo.transaction.receiver_id,
+    });
+  }
+
+  const gasUsed = getGasUsed(
+    transactionInfo.transaction_outcome,
+    receiptsOutcome
+  );
+  const txActions =
+    transactionInfo.transaction.actions.map(mapRpcActionToAction);
+  const gasAttached = getGasAttached(gasUsed, txActions);
+  const transactionFee = getTransactionFee(
+    transactionInfo.transaction_outcome,
+    receiptsOutcome
+  );
+  const transactionAmount = getDeposit(txActions);
+
+  return {
+    hash: transactionHash,
+    created: {
+      timestamp: transactionBaseInfo.blockTimestamp,
+      blockHash: transactionBaseInfo.blockHash,
+    },
+    signerId: transactionInfo.transaction.signer_id,
+    receiverId: transactionInfo.transaction.receiver_id,
+    index: transactionBaseInfo.transactionIndex,
+    fee: transactionFee,
+    amount: transactionAmount,
+    status: convertRpcTxStatusToStatus(transactionInfo.status),
+    gasUsed,
+    gasAttached,
+  };
+};
+
+const convertRpcTxStatusToStatus = (
+  status: RPC.FinalExecutionStatus
+): TransactionStatus => {
+  const txStatus = Object.keys(status)[0];
+  switch (txStatus) {
```

Comment:
So what's the state machine of those statuses?
Not Started -> Started -> Fetching -> Either(Failure, Success)?

Reply:
Yes. Not Started -> Started -> Failure || Success

Reply:
You will never receive "Not Started" or "Started" from RPC, so it should be just "Failure" or "Success" for now. "Not Started" and "Started" statuses were added there due to my lack of understanding of this fact back in the days. We may add support for incomplete transactions in the future, but it is not that crucial given how fast NEAR network is.

Reply:
So it should be something like this now?
```
switch (txStatus) {
  case "Failure":
    return "fail";
  case "SuccessValue":
    return "success";
  default:
    return "fetching";
}
```